File: python/hardware/expansion/expansion_card.py
from datetime import date
from components.component import Component
from components.hardware import Hardware
from interfaces.connectable import Connectable
from interfaces.upgradable import Upgradable

class ExpansionCard(Hardware, Component, Connectable, Upgradable):
    def __init__(self, interface_type="", serial_number="", installation_date=None, physically_installed=False, id="", manufacturer="", model="", price=0.0):
        Hardware.__init__(self, serial_number, installation_date, physically_installed)
        Component.__init__(self, id, manufacturer, model, price)
        self._interface_type = interface_type

    # Setters
    def set_interface_type(self, interface_type):
        self._interface_type = interface_type

    # Getters
    def get_interface_type(self):
        return self._interface_type

    def connect(self):
        print("Expansion card connected")
        return True

    def disconnect(self):
        print("Expansion card disconnected")
        return True

    def upgrade(self):
        print("Expansion card upgraded")
        return True

    def downgrade(self):
        print("Expansion card downgraded")
        return True

File: cpp/hardware/expansion/ExpansionCard.cpp
#pragma once
#include <bits/stdc++.h>

#include "../../component/Component.cpp"
#include "../../component/Hardware.cpp"
#include "../../interfaces/Connectable.cpp"
#include "../../interfaces/Upgradable.cpp"

class ExpansionCard : public Hardware, public Connectable, public Component
{
protected:
    std::string interfaceType;

public:
    // Constructor
    ExpansionCard() : Hardware(), Component(), interfaceType("") {}
    ExpansionCard(const std::string &type, const std::string &serial, const Date &date, bool installed, const std::string &id, const std::string &manufacturer, const std::string &model, double price)
        : Hardware(serial, date, installed), Component(id, manufacturer, model, price), interfaceType(type) {}

    // Destructor
    ~ExpansionCard() {}

    // Getters
    std::string getInterfaceType() const
    {
        return interfaceType;
    }

    // Setters
    void setInterfaceType(const std::string &type)
    {
        interfaceType = type;
    }

    bool connect() override
    {
        printf("Expansion card connected\n");
        return true;
    }

    bool disconnect() override
    {
        printf("Expansion card disconnected\n");
        return true;
    }
};